source/common/http/utility.cc

```diff
@@ -441,6 +441,9 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   ResponseHeaderMapPtr response_headers{createHeaderMap<ResponseHeaderMapImpl>(
       {{Headers::get().Status, std::to_string(enumToInt(response_code))}})};
 
+  if (encode_functions.modify_headers_) {
+    encode_functions.modify_headers_(*response_headers);
+  }
```

Comment:
So the problem as I understand it, was that modify_headers_ is setting an incorrect content length, and then we need to run the

 if (!body_text.empty()) {
    response_headers->setContentLength(body_text.size());
    response_headers->setReferenceContentType(content_type);	
}
just before encoding to ensure that both values are set correctly.
Aren't we still not necessarily doing that correctly for gRPC trailer-only responses?  if the modify headers function or encode functions set content length based on the body, and then we hit the "Respond with a gRPC trailers-only response if the request is gRPC"  block below, I think we'd update the content type but not the content length.
If I'm reading it right, mind fixing that in this PR as well?

Reply:
We are doing that in the old code.  But encode_functions.encode_headers_  is calling `modify_headers` function that may override them.   That is why this change move `modify_headers` out of encode_functions.encode_headers_ and pass it to this function directly.

Reply:
For grpc trailer-only response, it should have the same issue if `modify_headers` is called in side encode_headers_ function.  By moving it out, it should solve the problem.

Reply:
Your third issue of if `modify_headers` is supposed to set content-type and content-length for the response body. My take is: for the requests using Utility::sendLocalReply, the response body is generated by this function, so content-type and length should be set by this function, it should not be set by `modify_headers`.  This change is correct to call `modify_headers` first, and override its content-type and length headers

Reply:
I fixed gRPC trailer-only issue:  remove content-length if modify_headers set it.